R/apply_tfrmt_methods.R

```diff
@@ -0,0 +1,240 @@
+    fmt_to_apply <- frmt_def$fmt_ls[[var]]
+    .data %>%
+      filter(!!param == var) %>%
+      apply_frmt(
+        frmt_def = fmt_to_apply,
+        .data = .,
+        values = values,
+        column = column,
+        param = param,
+        label = label,
+        group = group,
+        ...
+      )
+  })
+
+  #Test if common information exists
+  miss_param_from_data <- .tmp_data %>%
+    pull(!!param) %>%
+    unique() %>%
+    setdiff(fmt_param_vals, .)
+
+  if(length(miss_param_from_data) > 0 ){
+    stop(paste0("Unable to create formatting combination because the following parameters are missing from the data:\n ",
+                paste0(miss_param_from_data, collapse = " \n")))
+  }
+
+  .tmp_data_wide <- .tmp_data %>%
+    select(!!values, !!param, !!column, !!label, !!!group) %>%
+    pivot_wider(
+      values_from = !!values,
+      names_from = !!param
+    ) %>%
+    mutate(
+      .is_all_missing =  all_missing(fmt_param_vals,.)
+    )
+
+  if(is.null(frmt_def$missing)){
+    frmt_def$missing <- ""
+  }
+
+
+  ## if both params are missing, then drop in frmt definition missing value
+  ## otherwise concat the params
+  .tmp_data_fmted <- .tmp_data_wide %>%
+    mutate(
+      !!values := case_when(
+        .data$.is_all_missing ~ frmt_def$missing,
+        TRUE ~ str_glue(frmt_def$expression) %>% as.character()
+      )
+    ) %>%
+    select(-all_of(fmt_param_vals), -.data$.is_all_missing)
+
+  ## keep only the first case of param, and add the joined values
+  .data %>%
+    filter(!!param == fmt_param_vals[[1]]) %>%
+    select(-!!values) %>%
+    left_join(
+      .tmp_data_fmted,
+      by = map_chr(c(column, label, group), as_label)
+    )
+
+}
+
+#' Apply form for frmt_when formats
+#'
+#' @param frmt_def formatting to be applied
+#' @param .data data, but only what is getting changed
+#' @param values values symbol should only be one
+#' @param ... additional arguments for methods
+#'
+#' @return rounded and formatted df
+#' @export
+#' @importFrom rlang as_label f_rhs f_lhs parse_exprs eval_tidy
+#' @importFrom dplyr pull if_else mutate
+#' @importFrom purrr map map_chr
+#' @importFrom rlang :=
+apply_frmt.frmt_when <- function(frmt_def, .data, values, ...){
+  values_str <- as_label(values)
+  n <- length(frmt_def)
+
+  val_len <- length(pull(.data, !!values))
+  right <- frmt_def %>%
+    map(f_rhs) %>%
+    map(function(x) {
+      if("frmt" %in% class(x)){
+        out <- apply_frmt(x, .data, values, ...) %>% pull(!!values)
+      } else {
+        out <- rep(x, val_len)
+      }
+      out
+    })
+
+  left <- frmt_def %>%
+    map_chr(f_lhs) %>%
+    if_else(. == "TRUE", ., paste0(values_str, .)) %>%
+    parse_exprs() %>%
+    map(eval_tidy, .data)
+
+
+  out <- rep(NA_character_, val_len)
+  replaced <- rep(FALSE, val_len)
+
+  for(i in seq_len(n)){
+    out <- replace_val(out, left[[i]] & !replaced, right[[i]])
+    replaced <- replaced | (left[[i]] & !is.na(left[[i]]))
+  }
```

Comment:
Would it be less computationally expensive to sort this out first? then only apply the frmt's as necessary? Just a discussion point, not a need to change IMO. This seems to be how ifelse and case_when handles it usually - calculate the values and then drop them into a vector of expected outputs.

Reply:
I don't think there is much of a cost difference, given calculating frmt's is relatively cheep. But, yes we can keep this tag for refactor if speed becomes an issue